<img align="right" height="90" src="/doc/logo.png">

# Test
Below you'll find detailed information on running modPhred pipeline. 

- **[Download test data](#download-test-data)**
- **[Run basecalling](#run-basecalling)**  
- **[Run modPhred](#run-modphred)**
- **[Run megalodon](#run-megalodon)**
- **[Compare modPhred and megalodon results](#compare-modphred-and-megalodon-results)**
- **[Compare samples](#compare-samples)**
- **[Calculate and plot correlation between modified sites](#calculate-and-plot-correlation-between-modified-sites)**
- **[Test data generation (for the curious ones ;) )](#test-data-generation-for-the-curious-ones--)**  

## Download test data
Get raw Fast5 data from [PRJEB22772](https://www.ebi.ac.uk/ena/data/view/PRJEB22772) (subset of reads from `NC_000913.3:1-100000`):
```bash
mkdir -p test
cd test
wget https://public-docs.crg.es/enovoa/public/lpryszcz/src/modPhred/test/ -q --show-progress -r -c -nc -np -nH --cut-dirs=6 --reject="index.html*"
```

## Run basecalling
Below, will basecall all runs from PRJEB22772. This will take 3-6 minutes on modern GPU or ~13 hours on CPU. 
```bash
acc=PRJEB22772; ver=3.4.1
for f in $acc/*; do
 d=`echo $f | rev | cut -f1 -d'/' | rev`;
 if [ ! -d guppy$ver/$acc/$d ]; then
  echo `date` $f $d;
  ~/src/ont-guppy_$ver/bin/guppy_basecaller --device cuda:0 -c dna_r9.4.1_450bps_modbases_dam-dcm-cpg_hac.cfg --compress_fastq --fast5_out --disable_pings -ri $f -s guppy$ver/$acc/$d;
 fi;
done; date
```
Above will work only if you have Nvidia GPU and CUDA installed. Alternatively, you can use CPU basecalling by skipping `--device cuda:0` parameter.
Note basecalling on CPU will be muuuuch slower (~100-300 times) than GPU, depending on your CPU/GPU model!
For some stats, have a look in [Table 1 in our recent paper](https://www.biorxiv.org/content/10.1101/818336v1).
With guppy v3.4.1 [consider using `--num_callers 4` instead of default 1 to speed-up basecalling ~2x](https://twitter.com/lpryszcz/status/1204854476949139463).

### Download pre-basecalled reads (optionally)
If for some reason you can't run basecalling, you can download basecalled Fast5 files using:
```bash
wget https://public-docs.crg.es/enovoa/public/lpryszcz/src/modPhred/basecalled/ -q --show-progress -r -c -nc -np -nH --cut-dirs=6 --reject="index.html*"
```

## Run modPhred
Running entire modPhred pipeline (~4 minutes):
```bash
acc=PRJEB22772; ver=3.4.1
~/src/modPhred/run -f ref/ECOLI.fa -o modPhred/$acc -i guppy$ver/$acc/*/workspace -t3
```

Instead you can run all steps one by one as follows:
```bash
~/src/modPhred/src/mod_encode.py -i _archives/guppy$ver/$acc/*/workspace
~/src/modPhred/src/mod_align.py -f ref/ECOLI.fa -o modPhred/$acc -i _archives/guppy$ver/$acc/*/workspace
~/src/modPhred/src/mod_report.py -f ref/ECOLI.fa -o modPhred/$acc -i _archives/guppy$ver/$acc/*/workspace
~/src/modPhred/src/mod_plot.py -i modPhred/$acc/mod.gz
```

Some examples of visualisation of the obtained results are described [here](/doc#visualisation). 

## Run megalodon
This is optional step - do it only if you wish to compare results generated by modPhred
to those generated by ONT's megalodon. 
Below will run megalodon on all samples (~2.5 hours):
```bash
acc=PRJEB22772
for f in $acc/*; do
 d=`echo $f | rev | cut -f1 -d'/' | rev`;
 mkdir -p megalodon/$acc
 if [ ! -d megalodon/$acc/$d ]; then
  echo `date` $f $d;
  megalodon $acc/$d --reference ref/ECOLI.fa --output-directory megalodon/$acc/$d --outputs basecalls mappings mods --devices 0 --processes 3 --verbose-read-progress 3
 fi
done; date
```
If you don't have megalodon installed, you can get it [here](https://github.com/nanoporetech/megalodon).


### Downloading pre-computed results (optionally)
If you have some problems with modPhred or megalodon, you can download final results using
```bash
wget https://public-docs.crg.es/enovoa/public/lpryszcz/src/modPhred/final/  -q --show-progress -r -c -nc -np -nH --cut-dirs=6 --reject="index.html*"
```

## Compare modPhred and megalodon results

```bash
# separate modPhred predictions for 6mA and 5mC and filter to those with 5% frequency
for f in modPhred/PRJEB22772/minimap2/*.bam.bed; do echo $f; for m in 6mA 5mC; do grep -w $m $f | awk '$11>=5' > $f.$m.flt.bed; done; done

# filter results to only sites with at least 5% of reads carrying modification
for f in megalodon/PRJEB22772/*/modified_bases.???.bed; do echo $f; awk '$11>=5' $f > $f.flt.bed; done

# get number of predictions for each run
wc -l modPhred/PRJEB22772/minimap2/*.flt.bed megalodon/PRJEB22772/*/modified_bases*.flt.bed

# draw Venn diagram for both mods
for m in 6mA 5mC; do
 ~/src/modPhred/src/mod_plot.py --venn {modPhred,megalodon}/PRJEB22772/*/*$m*.flt.bed -n modPhred_1D modPhred_2D megalodon_1D megalodon_2D -o venn.$m.05.svg;
done
```

Above will produce Venn diagrams similar to these:  
<img align="right" width="400" src="/test/venn.6mA.05.svg">
<img width="400" src="/test/venn.5mC.05.svg">

As you can see, megalodon predicts way more modified sites than modPhred,
but those additional predictions 

For more detailed comparison have a look in [modPhred paper](/doc#citation).

## Compare samples
You can visualise depth of coverage, basecall accuracy, modification frequency and
median modification probability for all modifications, for all modified positions
between all samples using:

```bash
~/src/modPhred/src/mod_plot.py --scatter -i modPhred/PRJEB22772/mod.gz
```

Above will produce scatter (pairwise) plots similar to these:  
<img align="right" width="400" src="/test/depth.png">
<img width="400" src="/test/mod_frequency.png">

## Calculate and plot correlation between modified sites
You can plot correlation between modified positions within certain regions. 
(EXTEND). 
```bash
~/src/modPhred/src/mod_correlation.py -i modPhred/PRJEB22772/mod.gz -r NC_000913.3:1-5000
```
This will generate something like that:

<img src="/test/NC_000913.3:1-5000.csv.gz.svg">

In addition you can narrow those plots to:
- only modifications from particular strand
```bash
~/src/modPhred/src/mod_correlation.py -i modPhred/PRJEB22772/mod.gz -r NC_000913.3:1-5000+
```

- only one modification from particular strand
```bash
~/src/modPhred/src/mod_correlation.py -i modPhred/PRJEB22772/mod.gz -r NC_000913.3:1-5000+ --mod 6mA
```

resulting in something like those:

<img align="right" width="400" src="/test/NC_000913.3:1-5000+.csv.gz.svg">
<img width="400" src="/test/NC_000913.3:1-5000+.6mA.csv.gz.svg">

You can find more information regarding those plots [here](/doc/#Correlations).

## Test data generation (for the curious ones ;) )
The test data was generated from [PRJEB22772](https://www.ebi.ac.uk/ena/data/view/PRJEB22772) by selecting only reads from NC_000913.3:1-100000 as follows:  
```bash
acc=PRJEB22772
for d in _archives/raw/$acc/*; do
 s=`echo $d|rev|cut -f1 -d"/"|rev`
 echo `date` $d $s
 if [ ! -d ~/src/modPhred/test/$acc/$s ]; then
  mkdir -p ~/src/modPhred/test/$acc/$s
  # get read IDs
  samtools view modPhred/$acc/minimap2/$s.bam "NC_000913.3:1-100000" | cut -f1 | sort | uniq > modPhred/$acc/minimap2/$s.bam.ids
  # subset reads
  python ~/src/ont_fast5_api/ont_fast5_api/conversion_tools/multi_fast5_subset.py -t 4 --recursive -l modPhred/$acc/minimap2/$s.bam.ids -i $d -s ~/src/modPhred/test/$acc/$s
 fi
done
```

If you have any questions or comments please share them in [Issues section](issue).
Cheers! 